pkg/controller/direct/vmwareengine/externalaccessrule_controller.go

```diff
@@ -0,0 +1,290 @@
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type externalAccessRuleAdapter struct {
+	gcpClient *gcp.Client
+	id        *krm.ExternalAccessRuleIdentity
+	desired   *krm.VMwareEngineExternalAccessRule
+	actual    *pb.ExternalAccessRule
+	reader    client.Reader
+}
+
+var _ directbase.Adapter = &externalAccessRuleAdapter{}
+
+func (a *externalAccessRuleAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting vmwareengine external access rule", "name", a.id)
+
+	req := &pb.GetExternalAccessRuleRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetExternalAccessRule(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting vmwareengine external access rule %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *externalAccessRuleAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating vmwareengine external access rule", "name", a.id)
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := a.desired.DeepCopy()
+	resource := VMwareEngineExternalAccessRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &pb.CreateExternalAccessRuleRequest{
+		Parent:               a.id.Parent().String(),
+		ExternalAccessRuleId: a.id.ID(),
+		ExternalAccessRule:   resource,
+	}
+	op, err := a.gcpClient.CreateExternalAccessRule(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating vmwareengine external access rule %s: %w", a.id.String(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("vmwareengine external access rule %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created vmwareengine external access rule in gcp", "name", a.id)
+
+	status := &krm.VMwareEngineExternalAccessRuleStatus{}
+	status.ObservedState = VMwareEngineExternalAccessRuleObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *externalAccessRuleAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating vmwareengine external access rule", "name", a.id)
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := a.desired.DeepCopy()
+	resource := VMwareEngineExternalAccessRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	paths := []string{}
+	if desired.Spec.Description != nil && !reflect.DeepEqual(resource.Description, a.actual.Description) {
```

Comment:
Just curious, why do you prefer DeepEqual() here instead of CompareProtoMessage()?